### ISSUE TITLE
kube-proxy and kube-router get `Unauthorized` response when controller started as HA cluster of 3 docker containers

### ISSUE THREAD

**Issue by Friz-zy** [k0sproject/k0s#3120]
### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Platform

```shell
Linux docker-01 4.19.130-boot2docker #1 SMP Mon Jun 29 23:52:55 UTC 2020 x86_64 GNU/Linux
```

### Version

v1.26.4-k0s.0

### Sysinfo

bash-5.1# k0s sysinfo
Machine ID: "db977dc6d5356e22e20f59aef12d6074fa6ffad2e4f3a32a841b86803e5bef3a" (from machine) (pass)
Total memory: 3.9 GiB (pass)
Disk space available for /var/lib/k0s: 6.0 GiB (pass)
Operating system: Linux (pass)
  Linux kernel release: 4.19.130-boot2docker (pass)
  Max. file descriptors per process: current: 1048576 / max: 1048576 (pass)
  Executable in path: modprobe: /sbin/modprobe (pass)
  /proc file system: mounted (0x9fa0) (pass)
  Control Groups: version 1 (pass)
    cgroup controller "cpu": available (pass)
    cgroup controller "cpuacct": available (pass)
    cgroup controller "cpuset": available (pass)
    cgroup controller "memory": available (pass)
    cgroup controller "devices": available (pass)
    cgroup controller "freezer": available (pass)
    cgroup controller "pids": available (pass)
    cgroup controller "hugetlb": available (pass)
    cgroup controller "blkio": available (pass)
  CONFIG_CGROUPS: Control Group support: no kernel config found (warning)
  CONFIG_NAMESPACES: Namespaces support: no kernel config found (warning)
  CONFIG_NET: Networking support: no kernel config found (warning)
  CONFIG_EXT4_FS: The Extended 4 (ext4) filesystem: no kernel config found (warning)
  CONFIG_PROC_FS: /proc file system support: no kernel config found (warning)

### What happened?

I'm experimenting with running k0s in docker (compose\swarm). When I run controller (etcd cluster, api, etc...) as 3 containers (HA), kube-proxy and kube-router fail with `Unauthorized` error in logs:

proxy:
```
2023-05-14T11:45:05.342332355Z stderr F I0514 11:45:05.341999       1 server.go:432] "Using lenient decoding as strict decoding failed" err="strict decoding error: unknown field \"udpIdleTimeout\""
2023-05-14T11:45:05.34292694Z stderr F W0514 11:45:05.342778       1 feature_gate.go:241] Setting GA feature gate ServiceInternalTrafficPolicy=true. It will be removed in a future release.
2023-05-14T11:45:05.346564602Z stderr F W0514 11:45:05.346379       1 feature_gate.go:241] Setting GA feature gate ServiceInternalTrafficPolicy=true. It will be removed in a future release.
2023-05-14T11:45:05.36090528Z stderr F E0514 11:45:05.360779       1 node.go:152] Failed to retrieve node info: Unauthorized
2023-05-14T11:45:06.553780555Z stderr F E0514 11:45:06.553321       1 node.go:152] Failed to retrieve node info: Unauthorized
2023-05-14T11:45:08.628226862Z stderr F E0514 11:45:08.627524       1 node.go:152] Failed to retrieve node info: Unauthorized
2023-05-14T11:45:13.375514438Z stderr F E0514 11:45:13.372379       1 node.go:152] Failed to retrieve node info: Unauthorized
2023-05-14T11:45:21.931378176Z stderr F E0514 11:45:21.930661       1 node.go:152] Failed to retrieve node info: Unauthorized
2023-05-14T11:45:38.24794541Z stderr F E0514 11:45:38.235975       1 node.go:152] Failed to retrieve node info: Unauthorized
***
2023-05-14T11:50:26.331736882Z stderr F E0514 11:50:26.331432       1 reflector.go:140] vendor/k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.Node: the server has asked for the client to provide credentials (get nodes)
2023-05-14T11:50:27.264754072Z stderr F W0514 11:50:27.264071       1 reflector.go:424] vendor/k8s.io/client-go/informers/factory.go:150: failed to list *v1.Node: Unauthorized
```

router:
```
***
2023-05-14T11:51:23.278530684Z stderr F I0514 11:51:23.274837       1 version.go:21] Running /usr/local/bin/kube-router version v1.5.1, built on 2022-07-29T22:25:31+0000, go1.17.10
2023-05-14T11:52:21.428767747Z stderr F W0514 11:52:21.428082       1 reflector.go:324] pkg/mod/k8s.io/client-go@v0.23.7/tools/cache/reflector.go:167: failed to list *v1.Service: Unauthorized
2023-05-14T11:52:21.428809892Z stderr F E0514 11:52:21.428164       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.23.7/tools/cache/reflector.go:167: Failed to watch *v1.Service: failed to list *v1.Service: Unauthorized
2023-05-14T11:52:23.27866936Z stderr F failed to run kube-router: Failed to synchronize cache: 1m0s timeout
```

I tried static and dynamic k0s configuration, also I got similar problems before with v1.27.1-k0s.0.

When I run the same controller setup as one container - all works fine, like in your docker demo.

my controller compose file: https://github.com/tldr-devops/k0s-in-docker/blob/main/controller.yml

### Steps to reproduce

1. `git clone https://github.com/tldr-devops/k0s-in-docker && cd k0s-in-docker`
2. `docker-compose -f generate-secrets.yml up`
3. `docker-compose -f controller.yml up -d`
Wait untill all k0s containers will have status `(healthy)`
`docker-compose -f controller.yml ps`
4. `docker-compose -f kubelet.yml up -d`
Wait for containers creation
5. check pods: `docker-compose -f controller.yml exec k0s-1 k0s kubectl get all -A`

### Expected behavior

proxy and router works as expected (like with 1 controller container)

proxy:
```
2023-05-14T11:12:44.255528789Z stderr F I0514 11:12:44.255282       1 server.go:432] "Using lenient decoding as strict decoding failed" err="strict decoding error: unknown field \"udpIdleTimeout\""
2023-05-14T11:12:44.255617907Z stderr F W0514 11:12:44.255541       1 feature_gate.go:241] Setting GA feature gate ServiceInternalTrafficPolicy=true. It will be removed in a future release.
2023-05-14T11:12:44.255820175Z stderr F W0514 11:12:44.255687       1 feature_gate.go:241] Setting GA feature gate ServiceInternalTrafficPolicy=true. It will be removed in a future release.
2023-05-14T11:12:44.271832049Z stderr F I0514 11:12:44.271735       1 node.go:163] Successfully retrieved node IP: 10.0.2.15
2023-05-14T11:12:44.27211549Z stderr F I0514 11:12:44.272000       1 server_others.go:109] "Detected node IP" address="10.0.2.15"
2023-05-14T11:12:44.320289145Z stderr F I0514 11:12:44.320138       1 server_others.go:170] "kube-proxy running in single-stack mode, this ipFamily is not supported" ipFamily=IPv6
2023-05-14T11:12:44.320365099Z stderr F I0514 11:12:44.320225       1 server_others.go:176] "Using iptables Proxier"
2023-05-14T11:12:44.320370924Z stderr F I0514 11:12:44.320278       1 proxier.go:242] "Setting route_localnet=1 to allow node-ports on localhost; to change this either disable iptables.localhostNodePorts (--iptables-localhost-nodeports) or set nodePortAddresses (--nodeport-addresses) to filter loopback addresses"
2023-05-14T11:12:44.321193125Z stderr F I0514 11:12:44.321086       1 server.go:655] "Version info" version="v1.26.4"
2023-05-14T11:12:44.321201721Z stderr F I0514 11:12:44.321114       1 server.go:657] "Golang settings" GOGC="" GOMAXPROCS="" GOTRACEBACK=""
2023-05-14T11:12:44.322357227Z stderr F I0514 11:12:44.321908       1 config.go:317] "Starting service config controller"
2023-05-14T11:12:44.32237471Z stderr F I0514 11:12:44.321940       1 shared_informer.go:270] Waiting for caches to sync for service config
2023-05-14T11:12:44.322378271Z stderr F I0514 11:12:44.321966       1 config.go:226] "Starting endpoint slice config controller"
2023-05-14T11:12:44.322393467Z stderr F I0514 11:12:44.321974       1 shared_informer.go:270] Waiting for caches to sync for endpoint slice config
2023-05-14T11:12:44.326139309Z stderr F I0514 11:12:44.324212       1 config.go:444] "Starting node config controller"
2023-05-14T11:12:44.326151587Z stderr F I0514 11:12:44.324239       1 shared_informer.go:270] Waiting for caches to sync for node config
2023-05-14T11:12:44.422931584Z stderr F I0514 11:12:44.422482       1 shared_informer.go:277] Caches are synced for endpoint slice config
2023-05-14T11:12:44.422974071Z stderr F I0514 11:12:44.422291       1 shared_informer.go:277] Caches are synced for service config
2023-05-14T11:12:44.425875281Z stderr F I0514 11:12:44.425052       1 shared_informer.go:277] Caches are synced for node config
```

router
```
***
2023-05-14T11:14:54.613992003Z stderr F -A KUBE-ROUTER-OUTPUT -m comment --comment "explicitly ACCEPT traffic that complies with network policies" -m mark --mark 0x20000/0x20000 -j ACCEPT
2023-05-14T11:14:54.614000384Z stderr F COMMIT
```

**Comment by Friz-zy** [comment]
All controllers share one CA for main pki and etcd + the same k0s config, all other files and configs are generated by k0s controller itself. For kubelet joining I tried both pregenerated and runtime tokens

**Comment by jnummelin** [comment]
You need to also share the Service Account token signing key with the controllers. Now each controller creates their own SA key pair which means only the controller who has signed the key will accept it, all others will refuse it.

When k0s generates it by itself, it uses 2048bit RSA key.

**Comment by Friz-zy** [comment]
@jnummelin thank you are lot! Could you tell me, if there are any other files that should be shared across controllers?

Imo this info should be added into docs about HA setup)
And also I have a proposal: would be very useful if k0s config will generate all necessary configs by some subcommand, maybe similar to [kubeadm certs](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/)

**Comment by jnummelin** [comment]
> Could you tell me, if there are any other files that should be shared across controllers?

The only common certs and key are for CA and SA signing, so essentially `ca.crt`, `ca.key`, `sa.key` and `sa.pub`.

> Imo this info should be added into docs about HA setup)

Yes, definitely would be good addition. Maybe see which part of the docs would make the most sense from your point of view and submit a PR?

> would be very useful if k0s config will generate all necessary configs 

Not really sure I get what you mean with this. 🤔 You mean k0s could have some sub-command to generate all the certs? Or to renew them? We actually to renew them automatically when k0s restarts. So it behaves quite similarly with kubeadm. So as long as you restart (usually with happens on an update) k0s once per year, your certs will stay valid.

**Comment by Friz-zy** [comment]
@jnummelin I'm sorry for delay. Yeah, I think that [custom-ca](https://docs.k0sproject.io/v1.27.2+k0s.0/custom-ca/) and [high-availability](https://docs.k0sproject.io/v1.27.2+k0s.0/high-availability/) should be updated. I'll try to write something good and create PR)

About sub-command for config: my bad, I didn't explain detailed my idea. Renew certs on each restart is a cool feature (btw also should be in docs somewhere), but I mean generation of all configs or at least pki one. For example, I want to pregenerate CA certs and other configs for share it between control node. I have 2 options: use external tools like openssl ([custom-ca](https://docs.k0sproject.io/v1.27.2+k0s.0/custom-ca/)) or use k0sctl, which one can't generate configs without ssh connection to hosts if I understand it right. k0s control plane generate all necessary certs and configs from one k0s config, but only during startup. Imo would be cool to have an option that I can execute apart for just generating all necessary pki and configs for whole cluster, and then distribute it without k0sctl)

**Comment by jnummelin** [comment]
@Friz-zy any chance you could write this up? :pretty-please: 😄

**Comment by Friz-zy** [comment]
@jnummelin done, check it please)

**Comment by jnummelin** [comment]
can be closed via #3245
